packages/angular_devkit/build_angular/src/webpack/hmr.js

```diff
@@ -0,0 +1,174 @@
+        parentNode.insertBefore(
+          document.createElement(element.tagName),
+          element,
+        );
+
+        parentNode.removeChild(element);
+      });
+
+    // Destroy old application, injectors, <style..., etc..
+    const platformRef = getPlatformRef(appRoot);
+    if (platformRef) {
+      platformRef.destroy();
+    }
+
+    // Restore all inputs and options
+    const bodyElement = document.querySelector('body');
+    if ((oldInputs.length + oldOptions.length) === 0 || !bodyElement) {
+      return;
+    }
+
+    // Use a `MutationObserver` to wait until the app-root element has been bootstrapped.
+    // ie: when the ng-version attribute is added.
+    new MutationObserver((_mutationsList, observer) => {
+      observer.disconnect();
+
+      const newAppRoot = getAppRoot();
+      if (!newAppRoot) {
+        return;
+      }
+      
+      const newAppRef = getApplicationRef(newAppRoot);
+      if (!newAppRef) {
+        return;
+      }
+
+      // Wait until the application isStable to restore the form values
+      newAppRef.isStable
+        .pipe(
+          filter(isStable => !!isStable),
+          take(1),
+        )
+        .subscribe(() => restoreFormValues(oldInputs, oldOptions));
+    })
+    .observe(bodyElement, {
+      attributes: true,
+      subtree: true,
+      attributeFilter: ['ng-version'],
+    });
+  });
+}
+
+function getAppRoot() {
+  const appRoot = document.querySelector('[ng-version]');
+  if (!appRoot) {
+    console.warn('[NG HMR] Cannot find the application root component.');
+
+    return undefined;
+  }
+
+  return appRoot;
+}
+
+function getToken(appRoot, token) {
+  return typeof ng === 'object' && ng.getInjector(appRoot).get(token) || undefined;
+}
+
+function getApplicationRef(appRoot) {
+  const appRef = getToken(appRoot, ApplicationRef);
+  if (!appRef) {
+    console.warn(`[NG HMR] Cannot get 'ApplicationRef'.`);
+
+    return undefined;
+  }
+
+  return appRef;
+}
+
+function getPlatformRef(appRoot) {
+  const platformRef = getToken(appRoot, PlatformRef);
+  if (!platformRef) {
+    console.warn(`[NG HMR] Cannot get 'PlatformRef'.`);
+
+    return undefined;
+  }
+
+  return platformRef;
+}
+
+function dispatchKeyUpEvent(element) {
+  element.dispatchEvent(new KeyboardEvent('keyup', { key: 'Enter' }));
+}
+
+function restoreFormValues(oldInputs, oldOptions) {
+  // Restore input
+  const newInputs = document.querySelectorAll('input');
+  if (newInputs.length && newInputs.length === oldInputs.length) {
+    console.log('[NG HMR] Restoring input values.');
+    for (let index = 0; index < newInputs.length; index++) {
+      const newElement = newInputs[index];
+      const oldElement = oldInputs[index];
+
+      switch (oldElement.type) {
+        case 'radio':
+        case 'checkbox':
+          newElement.checked = oldElement.checked;
+          break;
+        default:
+          newElement.value = oldElement.value;
+          break;
```

Comment:
Assuming the default is has a `value` prop could fail if W3C introduces a new `<input />` type that doesn't use `value`. I'm thinking we should list out all the currently supported types explicitly and then default to `console.warn("Unsupported input type.")` or something like that. It would be safer and avoid breaking us from otherwise non-breaking platform changes.

Reply:
Agree